staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go

```diff
@@ -0,0 +1,63 @@
+package cacher
+
+import (
+	"fmt"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+)
+
+type clusterNameAndNsGetter interface {
+	GetClusterName() string
+	GetNamespace() string
+}
+
+// recalculateWatchCacheKey adjusts the key under which the given object will be stored in the watch cache.
+//
+// resources in the db are stored without the clusterName, since the reflector used by the cache uses the logicalcluster.Wildcard
+// the clusterName will be assigned to object by the storage layer upon retrieval.
+// we need take it into consideration and change the key to contain the clusterName
+// because this is how clients are going to be retrieving data from the cache.
+func recalculateWatchCacheKey(existingKey string, object runtime.Object) string {
+	var clusterName, ns string
+	key := existingKey
+
+	switch t := object.(type) {
+	case metav1.ObjectMetaAccessor:
+		clusterName = t.GetObjectMeta().GetClusterName()
+		ns = t.GetObjectMeta().GetNamespace()
+	case clusterNameAndNsGetter:
+		clusterName = t.GetClusterName()
+		ns = t.GetNamespace()
+	default:
+		klog.Warningf("unknown object, could not get a clusterName and a namespace from: %T", object)
+		return key
+	}
+
+	if len(clusterName) == 0 {
+		klog.Warningf("no clusterName was found for: %T", object)
+		return key
+	}
+
+	// if we are dealing with a scoped resource put a clusterName before the namespace
+	// for example:
+	//
+	//  /group/resource/namespace/instance => /group/resource/clusterName/namespace/instance
+	//
+	//  /group/resource/$identity/namespace/instance => /group/resource/$identity/clusterName/namespace/instance
+	if len(ns) > 0 {
```

Comment:
To what extent do we need to control edge case behavior here? We haven't checked that the `ns` exists in the key yet, nor that it is at the 2nd position by path separators